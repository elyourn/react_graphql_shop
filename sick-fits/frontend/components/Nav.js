import React from 'react';
import { Mutation } from 'react-apollo';
import { TOGGLE_CART_MUTATION } from './Cart';

import Link from 'next/link';
import NavStyles from './styles/NavStyles';
import User from './User';
import SignOut from '../components/SignOut';
import CartCount from './CartCount';


const Nav = () => (
    <User>
        {({data: { me }}) => (
            <NavStyles>
                <Link href="/items">
                    <a>
                        Shop
                    </a>
                </Link>
                {me ? (
                    <React.Fragment>
                        <Link href="/sell">
                            <a>
                                Sell
                            </a>
                        </Link>
                        <Link href="/orders">
                            <a>
                                Orders
                            </a>
                        </Link>
                        <Link href="/me">
                            <a>
                                Account
                            </a>
                        </Link>
                        <SignOut />
                        <Mutation
                            mutation={TOGGLE_CART_MUTATION}
                        >
                            {(toggleCart) => (
                                <button onClick={toggleCart}>
                                    My cart
                                    <CartCount
                                        count={
                                            me.cart.reduce((tally, cartItem) => (tally + cartItem.quantity), 0)
                                        } 
                                    />
                                </button>
                            )}
                        </Mutation>
                        
                    </React.Fragment>
                ) : 
                    <Link href="/signup">
                        <a>
                            Sign In
                        </a>
                    </Link>    
                }
            </NavStyles>
        )}
    </User>
);
export default Nav;