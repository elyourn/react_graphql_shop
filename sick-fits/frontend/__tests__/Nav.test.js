import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import Nav from '../components/Nav';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser, fakeCartItem } from '../lib/testUtils';

const notSignedInMocks = [{
    request: { query: CURRENT_USER_QUERY },
    result: { data: {
        me: null,
    }},
}];

const signedInMocks = [{
    request: { query: CURRENT_USER_QUERY },
    result: { data: {
        me: fakeUser(),
    }},
}];

const signedInWithCartItemsMocks = [{
    request: { query: CURRENT_USER_QUERY },
    result: { data: {
        me: {
            ...fakeUser(), 
            cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()]
        },
    }},
}];

describe('<Nav />', () => {
    it('render a minimal nav when sign out', async () => {
        const wrapper = mount(
            <MockedProvider mocks={notSignedInMocks}>
                <Nav />
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        const nav = wrapper.find('NavStyles');
        expect(toJSON(nav)).toMatchSnapshot();
    });
    it('render a full nav when sign in', async () => {
        const wrapper = mount(
            <MockedProvider mocks={signedInMocks}>
                <Nav />
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        const nav = wrapper.find('NavStyles').children();
        expect(nav.children().length).toBe(6);
        expect(nav.text()).toContain('Sign out');
    });

    it('render the amount of items in the cart', async () => {
        const wrapper = mount(
            <MockedProvider mocks={signedInWithCartItemsMocks}>
                <Nav />
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        const nav = wrapper.find('NavStyles');

        const count = nav.find('div.count');
        expect(toJSON(count)).toMatchSnapshot();
    });
});