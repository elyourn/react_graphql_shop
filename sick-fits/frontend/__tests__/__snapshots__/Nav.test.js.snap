// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> render a minimal nav when sign out 1`] = `
<NavStyles>
  <ul
    className="NavStyles-sc-11c0d2g-0 caHxBV"
  >
    <Link
      href="/items"
    >
      <a
        href="/items"
        onClick={[Function]}
      >
        Shop
      </a>
    </Link>
    <Link
      href="/signup"
    >
      <a
        href="/signup"
        onClick={[Function]}
      >
        Sign In
      </a>
    </Link>
  </ul>
</NavStyles>
`;

exports[`<Nav /> render the amount of items in the cart 1`] = `
<div
  className="count CartCount__Dot-xxvp4g-1 fJsVOg"
>
  9
</div>
`;
